lib/Service/AuthenticationService.php

```diff
@@ -3,9 +3,138 @@
 namespace OCA\OpenConnector\Service;
 
 
+use GuzzleHttp\Client;
+use OAuthException;
+use Symfony\Component\HttpFoundation\Exception\BadRequestException;
 
 class AuthenticationService
 {
-    
+	public const REQUIRED_PARAMETERS_CLIENT_CREDENTIALS = [
+		'grant_type',
+		'scope',
+		'authentication',
+		'client_id',
+		'client_secret',
+	];
+	public const REQUIRED_PARAMETERS_PASSWORD = [
+		'grant_type',
+		'scope',
+		'authentication',
+		'username',
+		'password',
+	];
+
+	/**
+	 * Create call options for OAuth with Client Credentials
+	 *
+	 * @param array $configuration Configuration array for authentication.
+	 *
+	 * @return array|array[] The call options for OAuth with Client Credentials.
+	 */
+	private function createClientCredentialConfig(array $configuration): array
+	{
+		if(($diff = array_diff(self::REQUIRED_PARAMETERS_CLIENT_CREDENTIALS, array_keys(array: $configuration))) !== []) {
+			throw new BadRequestException(message: 'Some required parameters are not set: ['.implode(separator: ',', array: $diff).']');
+		}
+
+		$callConfig = [
+			'form_params' => [
+				'grant_type' => $configuration['grant_type'],
+				'scope'		 => $configuration['scope'],
+			]
+		];
+
+		if($configuration['authentication'] === 'body') {
+			$callConfig['form_params']['client_id']     = $configuration['client_id'];
+			$callConfig['form_params']['client_secret'] = $configuration['client_secret'];
+		} else if ($configuration['authentication'] === 'basic_auth') {
+			$callConfig['auth'] = [
+				'username' => $configuration['client_id'],
+				'password' => $configuration['client_secret'],
+			];
+		}
+		//@todo: check for off-cases, i.e. camelCase (not according to OAuth standards)
+
+		return $callConfig;
+	}
+
+	/**
+	 * Create call options for OAuth with Password Credentials
+	 *
+	 * @param array $configuration Configuration array for authentication.
+	 *
+	 * @return array|array[] The call options for OAuth with Password Credentials
+	 */
+	private function createPasswordConfig(array $configuration): array
+	{
+
+		if(($diff = array_diff(self::REQUIRED_PARAMETERS_PASSWORD, array_keys(array: $configuration))) !== []) {
+			throw new BadRequestException(message: 'Some required parameters are not set: ['.implode(separator: ',', array: $diff).']');
```

Comment:
Can the implode be assigned to a var on an empty line above?

Reply:
We could try that, yes